.devcontainer/launch.sh

```diff
@@ -59,20 +93,220 @@ parse_options() {
+    )"
+
+    if test -z "${REMOTE_USER:-}"; then
+        REMOTE_USER="$(
+            # Read remoteUser from image metadata
+            read_docker_image_metadata \
+          | json_string "remoteUser"
+        )"
+    fi
+
+    # Read runArgs
+    local -a RUN_ARGS="($(
+        json_array "runArgs" < "${path}/devcontainer.json"
+    ))"
+
+    # Read initializeCommand
+    local -a INITIALIZE_COMMAND="($(
+        json_array "initializeCommand" < "${path}/devcontainer.json"
+    ))"
+
+    local -a ENV_VARS="($(
+        json_map "containerEnv" < "${path}/devcontainer.json" \
+      | sed -r 's/(.*)=(.*)/--env \1=\2/'
+    ))";
+
+    local -a MOUNTS="($(
+        tee < "${path}/devcontainer.json"   \
+            1>/dev/null                     \
+            >(json_array "mounts")          \
+            >(json_string "workspaceMount") \
+      | xargs -r -I% echo --mount '%'
+    ))"
+
+    # Update run args and env vars
+
+    # Don't pass -it if running in CI
+    if [ "${CI:-false}" != "true" ]; then
+        RUN_ARGS+=("-it")
+    fi
+
+    for flag in rm init; do
+        if [[ " ${RUN_ARGS[*]} " != *" --${flag} "* ]]; then
+            RUN_ARGS+=("--${flag}")
+        fi
+    done
+
+    if test -n "${gpu_request:-}"; then
+        RUN_ARGS+=(--gpus "${gpu_request}")
+    else
+        # Read hostRequirements.gpu
+        local GPU_REQUEST="$(
+            json_map "hostRequirements" < "${path}/devcontainer.json" \
+          | grep 'gpu=' \
+          | sed -r 's/(.*)=(.*)/\2/' \
+          | xargs
+        )"
+        if test "${GPU_REQUEST:-false}" = true; then
+            RUN_ARGS+=(--gpus all)
+        elif test "${GPU_REQUEST:-false}" = optional && \
+             command -v nvidia-container-runtime >/dev/null 2>&1; then
+            RUN_ARGS+=(--gpus all)
+        fi
+    fi
+
+    RUN_ARGS+=(--workdir "${WORKSPACE_FOLDER:-/home/coder/cccl}")
+
+    if test -n "${REMOTE_USER:-}"; then
+        ENV_VARS+=(--env NEW_UID="$(id -u)")
+        ENV_VARS+=(--env NEW_GID="$(id -g)")
+        ENV_VARS+=(--env REMOTE_USER="$REMOTE_USER")
+        RUN_ARGS+=(-u root:root)
+        RUN_ARGS+=(--entrypoint "${WORKSPACE_FOLDER:-/home/coder/cccl}/.devcontainer/docker-entrypoint.sh")
+    fi
+
+    if test -n "${SSH_AUTH_SOCK:-}"; then
+        ENV_VARS+=(--env "SSH_AUTH_SOCK=/tmp/ssh-auth-sock")
+        MOUNTS+=(--mount "source=${SSH_AUTH_SOCK},target=/tmp/ssh-auth-sock,type=bind")
+    fi
+
+    if test -v volumes && test ${#volumes[@]} -gt 0; then
+        MOUNTS+=("${volumes[@]}")
+    fi
+
+    if test -v env_vars && test ${#env_vars[@]} -gt 0; then
+        ENV_VARS+=("${env_vars[@]}")
+    fi
+
+    if test "${#INITIALIZE_COMMAND[@]}" -gt 0; then
+        eval "${INITIALIZE_COMMAND[*]@Q}"
+    fi
+
+    exec docker run \
+        "${RUN_ARGS[@]}" \
+        "${ENV_VARS[@]}" \
+        "${MOUNTS[@]}" \
+        "${DOCKER_IMAGE}" \
+        "$@"
 }
 
 launch_vscode() {
+    local -;
+    set -euo pipefail;
```

Comment:
question: We usually just set this at the top-level at the beginning of our scripts. Is there a reason to only enable this per-function?

Reply:
This is mostly reflexive at this point, but I do it defensively for two reasons:
1. Some functions need to enable/disable different shellopts depending on what they do
2. I can't assume someone won't source my scripts and invoke the functions directly, in which case I wouldn't want to change the shellopts in their session
    * Sometimes I write scripts that are intended to be both sourced and run directly, and this means I don't have to go back and think about un-setting the top-level shellopts:
    ```shell
      _do_my_thing { local -; set -e; echo "foo"; }
      if [ "$(basename "${BASH_SOURCE[${#BASH_SOURCE[@]}-1]}")" = my-script.sh ]; then
        _do_my_thing "$@" <&0;
      fi
    ```